base/twiceprecision.jl

```diff
@@ -0,0 +1,508 @@
+# but not guaranteed to return a precise answer.
+# https://en.wikipedia.org/wiki/Continued_fraction#Best_rational_approximations
+function rat(x)
+    y = x
+    a = d = 1
+    b = c = 0
+    m = maxintfloat(narrow(typeof(x)))
+    while abs(y) <= m
+        f = trunc(Int,y)
+        y -= f
+        a, c = f*a + c, a
+        b, d = f*b + d, b
+        max(abs(a), abs(b)) <= convert(Int,m) || return c, d
+        oftype(x,a)/oftype(x,b) == x && break
+        y = inv(y)
+    end
+    return a, b
+end
+
+narrow(::Type{Float64}) = Float32
+narrow(::Type{Float32}) = Float16
+narrow(::Type{Float16}) = Float16
+
+function add2{T<:Number}(u::T, v::T)
+    @_inline_meta
+    u, v = ifelse(abs(v) > abs(u), (v, u), (u, v))
+    w = u + v
+    w, (u-w) + v
+end
+
+add2(u, v) = _add2(promote(u, v)...)
+_add2{T<:Number}(u::T, v::T) = add2(u, v)
+_add2(u, v) = error("$u::$(typeof(u)) and $v::$(typeof(v)) cannot be promoted to a common type")
+
+function +(x::TwicePrecision, y::Number)
+    s_hi, s_lo = add2(x.hi, y)
+    TwicePrecision(s_hi, s_lo+x.lo)
+end
++(x::Number, y::TwicePrecision) = y+x
+
+function +{T}(x::TwicePrecision{T}, y::TwicePrecision{T})
+    r = x.hi + y.hi
+    s = abs(x.hi) > abs(y.hi) ? (((x.hi - r) + y.hi) + y.lo) + x.lo : (((y.hi - r) + x.hi) + x.lo) + y.lo
+    TwicePrecision(r, s)
+end
++(x::TwicePrecision, y::TwicePrecision) = _add2(promote(x, y)...)
+_add2{T<:TwicePrecision}(x::T, y::T) = x + y
+_add2(x::TwicePrecision, y::TwicePrecision) = TwicePrecision(x.hi+y.hi, x.lo+y.lo)
+
+function *(x::TwicePrecision, v::Integer)
+    v == 0 && return TwicePrecision(x.hi*v, x.lo*v)
+    nb = ceil(Int, log2(abs(v)))
+    u = truncbits(x.hi, nb)
+    y_hi, y_lo = add2(u*v, ((x.hi-u) + x.lo)*v)
+    TwicePrecision(y_hi, y_lo)
+end
+
+function _mul2{T<:Union{Float16,Float32,Float64}}(x::TwicePrecision{T}, v::T)
+    v == 0 && return TwicePrecision(T(0), T(0))
+    xhh, xhl = splitprec(x.hi)
+    vh, vl = splitprec(v)
+    y_hi, y_lo = add2(xhh*vh, xhh*vl + xhl*vh)
+    TwicePrecision(y_hi, y_lo + xhl*vl + x.lo*v)
+end
+
+_mul2(x::TwicePrecision, v::Number) = TwicePrecision(x.hi*v, x.lo*v)
+
+function *{R,S<:Number}(x::TwicePrecision{R}, v::S)
+    T = promote_type(R, S)
+    _mul2(convert(TwicePrecision{T}, x), convert(T, v))
+end
+
+*(v::Number, x::TwicePrecision) = x*v
+
+function /(x::TwicePrecision, v::Number)
+    hi = x.hi/v
+    w = TwicePrecision(hi, zero(hi)) * v
+    lo = (((x.hi - w.hi) - w.lo) + x.lo)/v
+    y_hi, y_lo = add2(hi, lo)
+    TwicePrecision(y_hi, y_lo)
+end
+
+# hi-precision version of prod(num)/prod(den)
+# num and den are tuples to avoid risk of overflow
+function proddiv(T, num, den)
+    @_inline_meta
+    t = TwicePrecision(T(num[1]), zero(T))
+    t = _prod(t, tail(num)...)
+    _divt(t, den...)
+end
+function _prod(t::TwicePrecision, x, y...)
+    @_inline_meta
+    _prod(t * x, y...)
+end
+_prod(t::TwicePrecision) = t
+function _divt(t::TwicePrecision, x, y...)
+    @_inline_meta
+    _divt(t /  x, y...)
```

Comment:
super trivial, but if you end up making any additional revisions the extra space after the `/` here stuck out a bit